### PR TITLE
speedup detector drift detection with multiprocessing and enhance rotation center finder

### PR DESCRIPTION
**Pull request by KedoKudo**
Several new features specific to APS-1ID are introduced in this PR, including

- corr.py
    - implement a hybrid image pre-processor (denoise and contrast enhance) for data collected at APS-1ID

- alignment.py
    - use multiprocessing to speed up the detector drift detection
    - implement an adaptive search method of the slit-corners detection

- npmath.py
    - remove unused option

- rotation.py
    - allow users to skip the built-in smoothing when using   find_center_vo()
    - allow users to specify a guess center when using find_center_pc(), which is useful for cases with the rotation center close to one edge of the image.

### CODE REVIEW

File: tomopy/misc/npmath.py [hunk 1]
```diff
@@ -107,4 +107,4 @@ def pad(x): return np.hstack([x, np.ones((x.shape[0], 1))])
     #   therefore we need to transpose the matrix here
     #   to get the correct rotation
 
-    return scipy.linalg.lstsq(pad(pts_source), pad(pts_target), rcond=-1)[0].T
+    return scipy.linalg.lstsq(pad(pts_source), pad(pts_target))[0].T
```
Comment (dgursoy) [inline tomopy/misc/npmath.py]: Add newline.

File: tomopy/prep/alignment.py [hunk 2]
```diff
@@ -768,44 +784,127 @@ def detector_drift_adjust_aps_1id(imgstacks, slit_cnr_ref):
...
+                                'quadrant+',
+                                medfilt2_kernel_size, 
+                                medfilt_kernel_size,
+                               )
+    cnrs_found = np.array([quickDiff(proj_cnrs[n,:,:] - slit_cnr_ref) < 15 
+                            for n in nlist])
+    kernels = [(medfilt2_kernel_size+2*i, medfilt_kernel_size+2*j)
+                    for i in range(15)
+                    for j in range(15)]
+    counter = 0
```
Comment (dgursoy) [inline tomopy/prep/alignment.py]: Indentation fix.

File: tomopy/prep/alignment.py [hunk 3]
```diff
@@ -768,44 +784,127 @@ def detector_drift_adjust_aps_1id(imgstacks, slit_cnr_ref):
...
+            for idx, n_img in enumerate(nlist):
+                proj_cnrs[n_img,:,:] = slit_cnr_ref
+            break
+        else:
+            # test with differnt 2D and 1D kernels
+            ks2d, ks1d = kernels[counter]
 
-    # -- init affine transformation matrix
+        _cnrs = _calc_proj_cnrs(imgstacks, ncore, nlist, 'quadrant+',ks2d, ks1d)
+        for idx, _cnr in enumerate(_cnrs):
```
Comment (dgursoy) [inline tomopy/prep/alignment.py]: Blank after comma.

File: tomopy/prep/alignment.py [hunk 4]
```diff
@@ -768,44 +784,127 @@ def detector_drift_adjust_aps_1id(imgstacks, slit_cnr_ref):
...
+            n_img = nlist[idx]
+            cnr   = proj_cnrs[n_img,:,:]  # previous results
+            # NOTE:
+            #  The detector corner should not be far away from reference
+            #  -> adiff < 15
+            #  The detected corner should be stable
+            #  -> rdiff < 0.1 (pixel)s
+            adiff = quickDiff(_cnr - slit_cnr_ref)
+            rdiff = quickDiff(_cnr - cnr) 
+            if rdiff < 0.1 and adiff < 15:
```
Comment (dgursoy) [inline tomopy/prep/alignment.py]: quickDiff -> quick_diff?

File: tomopy/misc/corr.py [hunk 5]
```diff
@@ -665,3 +668,80 @@ def _get_mask(dx, dy, ratio):
         r2 = rad2 * rad2
     y, x = np.ogrid[0.5 - rad1:0.5 + rad1, 0.5 - rad2:0.5 + rad2]
     return x * x + y * y < ratio * ratio * r2
+
+
+def enhance_projs_aps_1id(imgstacks, median_ks=5, ncore=None):
+    """
+    Enhance the projection images with weak contrast collected at APS 1ID
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply tomopy/misc/corr.py]: I added additional explanations in the docstring for clarification.

File: tomopy/misc/corr.py [hunk 6]
```diff
@@ -665,3 +668,80 @@ def _get_mask(dx, dy, ratio):
         r2 = rad2 * rad2
     y, x = np.ogrid[0.5 - rad1:0.5 + rad1, 0.5 - rad2:0.5 + rad2]
     return x * x + y * y < ratio * ratio * r2
+
+
+def enhance_projs_aps_1id(imgstacks, median_ks=5, ncore=None):
+    """
```
[the remark this answers is not in the document]
Comment (KedoKudo) [reply tomopy/misc/corr.py]: I see your point here.  Singular form here makes much more sense.  I will refactor the function accordingly.

File: tomopy/prep/alignment.py (hunk 4 above)
Comment (KedoKudo) [reply tomopy/prep/alignment.py]: I updated the inline function name.

**Comment by dgursoy** [comment]
Thanks @KedoKudo.

**Pull request merged**